### PR TITLE
refactor: migrate vm with storage

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Refactors:
- `UserVmManagerImpl::migrateVirtualMachineWithVolume`
- `Boolean HypervisorCapabilitiesDao::isStorageMotionSupported` to `boolean HypervisorCapabilitiesDao::isStorageMotionSupported` for simplifying callers.
- `StorageManager::isStoragePoolComplaintWithStoragePolicy` to `StorageManager::isStoragePoolCompliantWithStoragePolicy`

Update log messages.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 1]
```diff
@@ -2393,7 +2393,7 @@ public boolean storagePoolHasEnoughSpaceForResize(StoragePool pool, long current
     }
 
     @Override
-    public boolean isStoragePoolComplaintWithStoragePolicy(List<Volume> volumes, StoragePool pool) throws StorageUnavailableException {
+    public boolean isStoragePoolCompliantWithStoragePolicy(List<Volume> volumes, StoragePool pool) throws StorageUnavailableException {
         if (volumes == null || volumes.isEmpty()) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:2397]: We can use `CollectionUtils` to do the validation.

```java
...
if (CollectionUtils.isEmpty(volumes)) {
...
```

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -6299,38 +6261,40 @@ public VirtualMachine migrateVirtualMachineWithVolume(Long vmId, Host destinatio
...
+
+        List<VolumeVO> vmVolumes = _volsDao.findUsableVolumesForInstance(vm.getId());
+        // Check if all the volumes are in the correct state.
+        for (VolumeVO volume : vmVolumes) {
+            if (volume.getState() != Volume.State.Ready) {
+                throw new CloudRuntimeException("Volume " + volume + " of the VM is not in Ready state. Cannot " + "migrate the vm with its volumes.");
             }
-            throw new InvalidParameterValueException("Migration of the vm " + vm + "from host " + srcHost + " to destination host " + destinationHost
-                    + " doesn't involve migrating the volumes.");
         }
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6298]: We could extract this code to a method with a properly name and remove the comment. Also, we could add unit tests to the method too.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -6365,24 +6329,75 @@ public VirtualMachine migrateVirtualMachineWithVolume(Long vmId, Host destinatio
...
+        if(serviceOfferingDetailsDao.findDetail(vm.getServiceOfferingId(), GPU.Keys.pciDevice.toString()) != null) {
+            throw new InvalidParameterValueException("Live Migration of GPU enabled VM is not supported");
+        }
+
+        // OfflineVmwareMigration: this condition is to complicated. (already a method somewhere)
+        if (!Arrays.asList(new HypervisorType[]{
+                HypervisorType.XenServer,
+                HypervisorType.VMware,
+                HypervisorType.KVM,
+                HypervisorType.Simulator}).contains(vm.getHypervisorType())) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6373]: This array will be created every time that this method is called, so we could extract this creation to a constant and use it here.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -6365,24 +6329,75 @@ public VirtualMachine migrateVirtualMachineWithVolume(Long vmId, Host destinatio
...
+            throw new InvalidParameterValueException(String.format("Unsupported hypervisor: %s for VM migration, we support XenServer/VMware/KVM only", vm.getHypervisorType()));
+        }
+
+        // Check that Vm does not have VM Snapshots
+        if (_vmSnapshotDao.findByVm(vmId).size() > 0) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6378]: We could assign the result of the method to a var and use `CollectionUtils` to validate it.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 5]
```diff
@@ -6365,24 +6329,75 @@ public VirtualMachine migrateVirtualMachineWithVolume(Long vmId, Host destinatio
...
+        // OfflineVmwareMigration: this condition is to complicated. (already a method somewhere)
+        if (!Arrays.asList(new HypervisorType[]{
+                HypervisorType.XenServer,
+                HypervisorType.VMware,
+                HypervisorType.KVM,
+                HypervisorType.Simulator}).contains(vm.getHypervisorType())) {
+            throw new InvalidParameterValueException(String.format("Unsupported hypervisor: %s for VM migration, we support XenServer/VMware/KVM only", vm.getHypervisorType()));
+        }
+
+        // Check that Vm does not have VM Snapshots
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6377]: As the code is clear and we have a explanation on the exception message, do still we need this comment?

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 4 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6383]: VMSnapshotDao::findByVm will always return a non-null list, I don't see an issue with directly checking the list size here.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@GutoVeronezi I resolved some of your comments on outdated code. Are all your questions/comments addressed (i.e. lgty?)

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 6]
```diff
@@ -575,6 +575,13 @@
     private static final ConfigKey<Boolean> VmDestroyForcestop = new ConfigKey<Boolean>("Advanced", Boolean.class, "vm.destroy.forcestop", "false",
             "On destroy, force-stop takes this value ", true);
 
+    public static final List<HypervisorType> VM_STORAGEMIGRATION_SUPPORTING_HYPERVISORS = new ArrayList<>(Arrays.asList(
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:578]: I think it would be better as `VM_STORAGE_MIGRATION_SUPPORTING_HYPERVISORS`.

**Review by GutoVeronezi — APPROVED** [review]
LGTM, I've did not test it though

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Pull request merged**